File: 139. Word Break/wordBreak.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
const wordBreak = (s, wordDict) => {
	const sLen = s.length;
  const dp = Array(sLen + 1).fill(false);
	dp[sLen] = true;

	for (let i = sLen; i >= 0; i--) {
		for (let j = 0; j < wordDict.length; j++) {
			if ((i + wordDict[j].length) <= sLen && s.substring(i, i +  wordDict[j].length) === wordDict[j]) {
				dp[i] = dp[i +  wordDict[j].length];
			}
			if (dp[i]) {
				break;
			}
		}
	}

	return dp[0];
};


const s = "leetcode";
const wordDict = ["leet", "code"];
console.log('wordBreak: ', wordBreak(s, wordDict))

File: 139. Word Break/wordBreak.py
from typing import List

class Solution:
  def wordBreak(self, s: str, wordDict: List[str]) -> bool:
    dp = [False] * (len(s) + 1)
    dp[len(s)] = True
    
    for i in range(len(s) - 1, -1, -1):
      print('i: ', i)
      for w in wordDict:
        if (i + len(w)) <= len(s) and s[i: i + len(w)] == w:
          dp[i] = dp[i + len(w)]
        if dp[i]:
          break
        
    return dp[0]
        
        
s = "leetcode"
wordDict = ["leet","code"]
solution = Solution()
print('wordBreak: ', solution.wordBreak(s, wordDict))
